test/ipns-pubsub.js

```diff
@@ -0,0 +1,172 @@
+
+const waitFor = require('./utils/wait-for')
+
+const config = {
+  Addresses: {
+    API: '/ip4/0.0.0.0/tcp/0',
+    Gateway: '/ip4/0.0.0.0/tcp/0',
+    Swarm: []
+  }
+}
+
+const namespace = '/record/'
+
+const spawnJsDaemon = (callback) => {
+  DaemonFactory.create({ type: 'js' })
+    .spawn({
+      disposable: true,
+      initOptions: { bits: 512 },
+      args: ['--enable-namesys-pubsub'], // enable ipns over pubsub
+      config
+    }, callback)
+}
+
+const spawnGoDaemon = (callback) => {
+  DaemonFactory.create()
+    .spawn({
+      disposable: true,
+      initOptions: { bits: 1024 },
+      args: ['--enable-namesys-pubsub'], // enable ipns over pubsub
+      config
+    }, callback)
+}
+
+const ipfsRef = '/ipfs/QmPFVLPmp9zv5Z5KUqLhe2EivAGccQW2r7M7jhVJGLZoZU'
+
+describe('ipns-pubsub', () => {
+  let nodeAId
+  let nodeBId
+  let nodes = []
+
+  // Spawn daemons
+  before(function (done) {
+    // CI takes longer to instantiate the daemon, so we need to increase the timeout
+    this.timeout(80 * 1000)
+
+    series([
+      (cb) => spawnGoDaemon(cb),
+      (cb) => spawnJsDaemon(cb)
+    ], (err, daemons) => {
+      expect(err).to.not.exist()
+      nodes = daemons
+
+      done()
+    })
+  })
+
+  // Get node ids
+  before(function (done) {
+    this.timeout(100 * 1000)
+
+    parallel([
+      (cb) => nodes[0].api.id(cb),
+      (cb) => nodes[1].api.id(cb)
+    ], (err, ids) => {
+      expect(err).to.not.exist()
+      expect(ids).to.exist()
+      expect(ids[0].id).to.exist()
+      expect(ids[1].id).to.exist()
+
+      nodeAId = ids[0]
+      nodeBId = ids[1]
+
+      nodes[0].api.swarm.connect(ids[1].addresses[0], () => {
+        setTimeout(done, 60000) // wait for republish as we can receive the republish message first
+      })
+    })
+  })
+
+  after(function (done) {
+    this.timeout(60 * 1000)
+    parallel(nodes.map((node) => (cb) => node.stop(cb)), done)
+  })
+
+  it('should get enabled state of pubsub', function (done) {
+    nodes[0].api.name.pubsub.state((err, state) => {
+      expect(err).to.not.exist()
+      expect(state).to.exist()
+      expect(state.enabled).to.equal(true)
+
+      done()
+    })
+  })
+
+  it('should publish the received record to a go node and a js subscriber should receive it', function (done) {
+    this.timeout(300 * 1000)
+
+    subscribeToReceiveByPubsub(nodes[0], nodes[1], nodeAId.id, done)
+  })
+
+  it('should publish the received record to a js node and a go subscriber should receive it', function (done) {
+    this.timeout(350 * 1000)
+
+    subscribeToReceiveByPubsub(nodes[1], nodes[0], nodeBId.id, done)
+  })
+})
+
+const subscribeToReceiveByPubsub = (nodeA, nodeB, id, callback) => {
+  let subscribed = false
+  function checkMessage (msg) {
+    subscribed = true
+  }
+
+  const keys = ipns.getIdKeys(fromB58String(id))
+  const topic = `${namespace}${base64url.encode(keys.routingKey.toBuffer())}`
+
+  nodeB.api.name.resolve(id, () => {
```

Comment:
This needs some explanation. What happens here? `nodeA` has not yet published anything for `self` - does it error or return `null`? We should handle the error at least.